packages/wallets/saifu/src/adapter.ts

```diff
@@ -0,0 +1,265 @@
+
+    private _connecting: boolean;
+    private _wallet: SaifuWallet | null;
+    private _publicKey: PublicKey | null;
+    private _readyState: WalletReadyState =
+        typeof window === 'undefined' || typeof document === 'undefined'
+            ? WalletReadyState.Unsupported
+            : WalletReadyState.NotDetected;
+
+    constructor(config: SaifuWalletAdapterConfig = {}) {
+        super();
+        this._connecting = false;
+        this._wallet = null;
+        this._publicKey = null;
+
+        if (this._readyState !== WalletReadyState.Unsupported) {
+            scopePollingDetectionStrategy(() => {
+                if (window.saifu) {
+                    this._readyState = WalletReadyState.Installed;
+                    this.emit('readyStateChange', this._readyState);
+                    return true;
+                }
+                return false;
+            });
+        }
+    }
+
+    get publicKey(): PublicKey | null {
+        return this._publicKey;
+    }
+
+    get connecting(): boolean {
+        return this._connecting;
+    }
+
+    get connected(): boolean {
+        return !!this._wallet?.isConnected;
+    }
+
+    get readyState(): WalletReadyState {
+        return this._readyState;
+    }
+
+    async connect(): Promise<void> {
+        try {
+            if (this.connected || this.connecting) return;
+            if (this._readyState !== WalletReadyState.Installed) throw new WalletNotReadyError();
+
+            this._connecting = true;
+
+            const wallet = window.saifu;
+
+            if (!wallet?.isConnected) {
+                try {
+                    // connection flow:
+                    // 1. Bind 'connect' for post auth
+                    // 2. Call wallet.connect()
+                    // 3. Wallet with authorize the user and emit 'connect'
+                    // 4. 'connect' will resolve the full promise
+                    await new Promise<void>((resolve, reject) => {
+
+                        // connect will get called post auth
+                        // resolving this promise will unblock the outer part
+                        // so this should happen last
+                        const connect = () => {
+                            wallet.off('connect', connect);
+                            resolve();
+                        };
+
+                        // saifu will emit 'connect' once connection has been established
+                        wallet.on('connect', connect);
+
+                        wallet.connect().catch((reason: any) => {
+                            wallet.off('connect', connect);
+                            reject(reason);
+                        });
+                    });
```

Comment:
This still contains Phantom hacks (using events to connect rather than Promise fulfillment). This only exists because Phantom doesn't always fulfill that Promise (for example, it doesn't reject when the extension window is closed). If `wallet.connect()` returns a Promise and is implemented properly, you don't need this entire block.

Reply:
The painpoints for imitating the most popular wallet with events. Okay let's do it the right way and handle the promise from `connect()` rather than events 👍 I removed the entire block and we'll migrate off it on the Saifu side in the next release as well